llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -18753,7 +18753,9 @@ SDValue DAGCombiner::rebuildSetCC(SDValue N) {
       EVT SetCCVT = N.getValueType();
       if (LegalTypes)
         SetCCVT = getSetCCResultType(SetCCVT);
-      // Replace the uses of XOR with SETCC
+      // Replace the uses of XOR with SETCC. Note, avoid this transformation if
+      // it would introduce illegal operations post-legalization as this can
+      // result in an infinite loop.
```

Comment:
"result in an infite loop" could use more details. " converting it to xor->setcc here, and expanding setcc->xor if the back-end requested it."

Speaking of which. I assume the infinite loop affects only targets that have `Expand` set on the op. Can we narrow the condition from `!Legal` to `==Expand` ?

Reply:
I can update to make the comment more descriptive but I'm not sure about narrowing the logic. `isCondCodeLegal` is used it lots of places to prevent combines that would produce illegal nodes and I'm not sure there would be a benefit to performing this combine if the resulting nodes are not legal. Further, Custom legalization or promotion could still maybe result in some sort of infinite loop. Is is alright if I leave the check as is?